b2share/modules/records/serializers/schemas/json.py

```diff
@@ -43,19 +47,35 @@ class DraftSchemaJSONV1(Schema):
     @pre_dump
     def filter_internal(self, data):
         """Remove internal fields from the record metadata."""
+        external_pids = []
+        bucket = None
+        try:
+            record = g.record
+            external_pids = generate_external_pids(record)
+            bucket = Bucket.query.filter_by(
+                id=str(record.files.bucket)).first()
+        except AttributeError:
+            record = None
+
         if '_deposit' in data['metadata']:
-            if 'external_pids' in data['metadata']['_deposit']:
-                data['metadata']['external_pids'] = \
-                    data['metadata']['_deposit']['external_pids']
             data['metadata']['owners'] = data['metadata']['_deposit']['owners']
+            # Add the external_pids only if the
+            # user is allowed to read the bucket
+            if external_pids and record and record.files:
+                if (allow_public_file_metadata(record) or
```

Comment:
We could check the permission before generating the list in order to avoid doing unnecessary queries to the DB.